pyadjoint/optimization/rol_solver_checkpointed.py

```diff
@@ -0,0 +1,283 @@
+
+        def __getstate__(self):
+            """Return a state tuple suitable for pickling"""
+
+            fname = os.path.join(
+                ROLCheckpointManager.get_checkpoint_dir(),
+                "vector_checkpoint_{}.h5".format(utils._new_uid()))
+            self.save(fname)
+
+            return (fname, self.inner_product)
+
+    class CheckPointedROLSolver(ROLSolver):
+        def __init__(self, problem, parameters, inner_product="L2"):
+            super().__init__(problem, parameters)
+            OptimizationSolver.__init__(self, problem, parameters)
+            self.rolobjective = ROLObjective(problem.reduced_functional)
+            x = [p.tape_value() for p in self.problem.reduced_functional.controls]
+            self.rolvector = CheckpointedROLVector(x, inner_product=inner_product)
+            self.params_dict = parameters
+
+            # self.bounds = super(CheckPointedROLSolver, self).__get_bounds()
+            # self.constraints = self.__get_constraints()
+
+    class LinMoreOptimiser(object):
+        def __init__(self, minimisation_problem, parameters, callback=None, checkpoint=False):
+
+            self.callback = callback
+            self.checkpoint_flag = checkpoint
+            if self.checkpoint_flag:
+                self.rol_solver = CheckPointedROLSolver(
+                    minimisation_problem, parameters, inner_product='L2')
+            else:
+                self.rol_solver = ROLSolver(
+                    minimisation_problem, parameters, inner_product='L2')
+
+            self.rol_parameters = ROL.ParameterList(
+                parameters, "Parameters")
+
+            self.rol_secant = ROL.InitBFGS(
+                parameters.get('General').get('Secant').get('Maximum Storage'))
+
+            self.rol_algorithm = ROL.LinMoreAlgorithm(
+                self.rol_parameters, self.rol_secant)
+
+            self.rol_algorithm.setStatusTest(
+                self.StatusTest(self.rol_parameters,
+                                self.rol_solver.rolvector,
+                                self),
+                False)
+
+        # solving the optimisation problem
+        def run(self):
+            self.rol_algorithm.run(
+                self.rol_solver.rolvector,
+                self.rol_solver.rolobjective,
+                self.rol_solver.bounds)
+
+        #
+        def checkpoint(self):
+
+            ROL.serialise_secant(self.rol_secant,
+                                 MPI.COMM_WORLD.rank,
+                                 ROLCheckpointManager.get_checkpoint_dir())
+
+            ROL.serialise_algorithm(self.rol_algorithm,
+                                    MPI.COMM_WORLD.rank,
+                                    ROLCheckpointManager.get_checkpoint_dir())
+
+            with CheckpointFile(os.path.join(
+                ROLCheckpointManager.get_checkpoint_dir(),
+                "solution_checkpoint.h5"),
+                    mode='w') as ckpnt:
```

Comment:
I think the way Black reformats this kind of statement is easier to read:

```suggestion
            with CheckpointFile(
                os.path.join(
                    ROLCheckpointManager.get_checkpoint_dir(), "solution_checkpoint.h5"
                ),
                mode="w",
            ) as ckpnt:
```